### PR TITLE
Separate the progress logging stream, use dyn trait instead of String for timestamps

### PR DESCRIPTION
**Pull request by utaal**
This only supports cfg(not(feature = "bincode")), bincode support upcoming.
Currently based on the `log_progress_accurately` branch.

@frankmcsherry

### CODE REVIEW

File: timely/src/logging.rs
Comment (frankmcsherry) [inline timely/src/logging.rs:78]: Just to check, `Timestamp` has already `Debug+Any+Abomonation/Serialize`. Is the motivation for the trait to add `Display`, or because the `communication::Data` trait isn't actually helpful?

**Comment by frankmcsherry** [comment]
Thanks very much!

I think this looks good, though I have to admit not knowing about ease/pain of logging out. Strings look pretty good, but other formats could be hard and I wouldn't know. This does the good things (imo) of extracting the progress logging to a new stream so that it can be turned off cheaply. I had a comment or two about the rationale for some traits, not because I doubt the rationale but more to write it down so that in a few months I can remember why it wasn't another way. :)

File: timely/src/logging.rs
Comment (utaal) [reply timely/src/logging.rs:78]: It’s because Data can’t be turned into a trait, because entomb takes a type parameter. Should probably add a comment to remind future readers.

**Comment by utaal** [comment]
I changed things to be serialization-agnostic, but to still offer the option of recovering the real timestamp type in the registered log function. This way, someone who needs detailed, typed information, can downcast to the concrete types of the timestamp present in the dataflow.

I also actually sent native types and not formatted strings, which I had failed to do previously (in `Progcaster`).

File: timely/src/logging.rs
Comment (frankmcsherry) [inline timely/src/logging.rs:76]: Random comment, not a change but: I bet this trait eventually becomes useful beyond progress events. At least, assuming there is any other reason to report information about timestamps (e.g., maybe other timestamp things, I dunno) they'll probably make use of this too. Again: nothing to change, just thinking out loud.

File: timely/src/logging.rs
Comment (frankmcsherry) [inline timely/src/logging.rs:109]: Minor, but this would be a great place to explain the need for a trait (vs a `type` or a `struct` that matches the same description).

File: timely/src/progress/broadcast.rs
Comment (frankmcsherry) [inline timely/src/progress/broadcast.rs:68]: I'm super confused that this works. I would have thought that as soon as this is boxed up you lose access to the `push` methods. Or is it because there is an auto-conversion from `Box<T>` to `Box<dyn Trait>` for any `T: Trait`? No change needed, just surprising to me!

File: timely/src/progress/broadcast.rs
Comment (frankmcsherry) [inline timely/src/progress/broadcast.rs:158]: Yeah, this is what I thought you would need to do. Is there a reason this is needed here but not previously? If so I think we should doc it, and if not we should unify up the two cases to look the same (maybe just aesthetics, but also to avoid randoms (myself) asking this question in the future).

**Review by frankmcsherry — APPROVED** [review]
I think this looks good. I have some questions, but none of them need to block merging this in to the PR (and potentially then merging the PR). I'll not merge right away in case any of the comments seem actionable to you, but otherwise I'm happy to do it now!

File: timely/src/logging.rs
Comment (utaal) [reply timely/src/logging.rs:76]: Yea, I thought the same when I wrote it, but it seemed unnecessary to push it to a separate module at this stage.

File: timely/src/progress/broadcast.rs
Comment (utaal) [reply timely/src/progress/broadcast.rs]: The asymmetry was unintentional (and I guess a potential perf defect). Fixed in the last commit.

File: timely/src/progress/broadcast.rs
Comment (utaal) [reply timely/src/progress/broadcast.rs:68]: I believe `push` works because of `https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-Deref` (`Deref` impl for `Box`). And yes, you can upcast a `Box<T>` to a `Box<dyn Trait>` for the same reason you can upcast a `&T` to a `&dyn Trait` for `T: Trait`.

Comment (utaal) [reply timely/src/progress/broadcast.rs:68]: To clarify, the type of `messages` and `internal` is still the concrete, non-`dyn` vector type. The upcast happens on assignment to the target struct below.

File: timely/src/logging.rs
Comment (utaal) [reply timely/src/logging.rs:109]: Added to the doc in the latest commit.

**Comment by utaal** [comment]
I believe I addressed the comments: so go ahead and merge if the latest changes look good! Thanks for the review.

**Pull request merged**
